src/scan/postgres_table_reader.cpp

```diff
@@ -0,0 +1,261 @@
+		bool interrupts_can_be_process = INTERRUPTS_CAN_BE_PROCESSED();
+
+		if (!interrupts_can_be_process) {
+			RESUME_CANCEL_INTERRUPTS();
+		}
+
+		if (!IsInParallelMode()) {
+			EnterParallelMode();
+			entered_parallel_mode = true;
+		}
+
+		ParallelContext *pcxt;
+		parallel_executor_info = PostgresFunctionGuard(ExecInitParallelPlan, table_scan_planstate,
+		                                               table_scan_query_desc->estate, nullptr, parallel_workers, -1);
+		pcxt = parallel_executor_info->pcxt;
+		PostgresFunctionGuard(LaunchParallelWorkers, pcxt);
+		nworkers_launched = pcxt->nworkers_launched;
+
+		if (pcxt->nworkers_launched > 0) {
+			PostgresFunctionGuard(ExecParallelCreateReaders, parallel_executor_info);
+			nreaders = pcxt->nworkers_launched;
+			parallel_worker_readers = (void **)palloc(nreaders * sizeof(TupleQueueReader *));
+			memcpy(parallel_worker_readers, parallel_executor_info->reader, nreaders * sizeof(TupleQueueReader *));
+		}
+
+		if (!interrupts_can_be_process) {
+			HOLD_CANCEL_INTERRUPTS();
+		}
+	}
+
+	elog(DEBUG1, "(PGDuckdDB/PostgresTableReader)\n\nQUERY: %s\nRUNNING: %s.\nEXECUTING: \n%s", table_scan_query,
+	     !nreaders ? "IN PROCESS THREAD" : psprintf("ON %d PARALLEL WORKER(S)", nreaders),
+	     ExplainScanPlan(table_scan_query_desc).c_str());
+
+	slot = PostgresFunctionGuard(ExecInitExtraTupleSlot, table_scan_query_desc->estate,
+	                             table_scan_planstate->ps_ResultTupleDesc, &TTSOpsMinimalTuple);
+}
+
+PostgresTableReader::~PostgresTableReader() {
+	std::lock_guard<std::mutex> lock(GlobalProcessLock::GetLock());
+
+	PostgresScopedStackReset scoped_stack_reset;
+
+	PostgresFunctionGuard(ExecEndNode, table_scan_planstate);
+
+	if (parallel_executor_info != NULL) {
+		PostgresFunctionGuard(ExecParallelFinish, parallel_executor_info);
+		PostgresFunctionGuard(ExecParallelCleanup, parallel_executor_info);
+	}
+
+	parallel_executor_info = NULL;
+
+	if (parallel_worker_readers) {
+		PostgresFunctionGuard(pfree, parallel_worker_readers);
+	}
+
+	parallel_worker_readers = NULL;
+
+	PostgresFunctionGuard(ExecutorFinish, table_scan_query_desc);
+	PostgresFunctionGuard(ExecutorEnd, table_scan_query_desc);
+	PostgresFunctionGuard(FreeQueryDesc, table_scan_query_desc);
+
+	if (entered_parallel_mode) {
+		ExitParallelMode();
+	}
+}
+
+int
+PostgresTableReader::ParallelWorkerNumber(Cardinality cardinality) {
+	static const int base_log = 8;
+	int cardinality_log = std::log2(cardinality);
+	int base = cardinality_log / base_log;
+	return std::max(1, std::min(base, std::max(max_workers_per_postgres_scan, max_parallel_workers)));
+}
+
+std::string
+PostgresTableReader::ExplainScanPlan(QueryDesc *query_desc) {
+	ExplainState *es = (ExplainState *)PostgresFunctionGuard(palloc0, sizeof(ExplainState));
+	es->str = makeStringInfo();
+	es->format = EXPLAIN_FORMAT_TEXT;
+	PostgresFunctionGuard(ExplainPrintPlan, es, query_desc);
+	std::string explain_scan(es->str->data);
+	return explain_scan;
+}
```

Comment:
Maybe we want to return a `const char*`? And as mentioned offline, either have caller wrap with `PostgresFunctionGuard` or introduce a new function and wrap it here?
```suggestion
const char*
PostgresTableReader::ExplainScanPlan_Unsafe(QueryDesc *query_desc) {
	ExplainState *es = (ExplainState *)palloc0(sizeof(ExplainState));
	es->str = makeStringInfo();
	es->format = EXPLAIN_FORMAT_TEXT;
	ExplainPrintPlan(es, query_desc);
	return es->str->data;
}

const char*
PostgresTableReader::ExplainScanPlan(QueryDesc *query_desc) {
	return PostgresFunctionGuard(ExplainScanPlan_Unsafe, query_desc);
}
```